### PR TITLE
Meta: Move all host tools to Lagom/Tools directory, reorganize CMake helpers

### PR DESCRIPTION
**Pull request by ADKaster**
First, move the CMake helpers in utils.cmake around and group the functions/settings into some more specific logical groupings.

I also moved the unicode_data.cmake helper into Meta/CMake, to centralize all our helper files.

And finally, the big one, move all our code generators and the configure components tool into a subdirectory of Meta/Lagom so that all targets that depend on LagomCore are in a subdirectory of Lagom. This lets us clean up the root CMakeLists.txt, and starts to make it look closer to how it'll need to look to complete #4594

### CODE REVIEW

File: Meta/CMake/utils.cmake [hunk 1]
```diff
@@ -1,3 +1,7 @@
+
+include(${CMAKE_CURRENT_LIST_DIR}/serenity_components.cmake)
+include(${CMAKE_CURRENT_LIST_DIR}/code_generators.cmake)
```
Comment (friendlyanon) [inline Meta/CMake/utils.cmake:3]: You can just use relative paths:
```suggestion
include(serenity_components.cmake)
include(code_generators.cmake)
```

File: CMakeLists.txt [hunk 2]
```diff
@@ -329,19 +318,22 @@ include_directories(${CMAKE_CURRENT_BINARY_DIR}/Userland/Services)
 include_directories(${CMAKE_CURRENT_BINARY_DIR}/Userland/Libraries)
 include_directories(${CMAKE_CURRENT_BINARY_DIR}/Userland)
 
+include(Meta/CMake/wasm_spec_tests.cmake)
+set(write_if_different ${SERENITY_PROJECT_ROOT}/Meta/write-only-on-difference.sh)
```
[the remark this answers is not in the document]
Comment (ADKaster) [reply CMakeLists.txt:322]: Uhh hmm. That does look to do the same thing that our little helper script does. @BenWiederhake  @nico  is there a technical reason we didn't use `cmake -E copy_if_different` when adding that script and the first few uses of it?

[the remark this answers is not in the document]
Comment (ADKaster) [reply CMakeLists.txt:322]: Oh I think our uses of that script would need to change a bit. the script can parse a bunch of arguments to create the file, and executes $2 with $3...N. 
So we have like
```
    add_custom_command(
        OUTPUT ${output}
        COMMAND ${write_if_different} ${output} $<TARGET_FILE:Lagom::IPCCompiler> ${source}
        VERBATIM
        DEPENDS Lagom::IPCCompiler
        MAIN_DEPENDENCY ${source}
    )
```
which would need to change to ... something else? To use ${CMAKE_COMMAND} -E copy_if_different, right? Unless there's an obvious extra argument to add_custom_command() that we could use.

[the remark this answers is not in the document]
Comment (friendlyanon) [reply CMakeLists.txt:322]: Your example could be rewritten as:
```cmake
add_custom_command(
    OUTPUT "${output}"
    BYPRODUCTS "${output}.tmp"
    COMMAND "$<TARGET_FILE:Lagom::IPCCompiler>" "${source}" > "${output}.tmp"
    COMMAND "${CMAKE_COMMAND}" -E copy_if_different "${output}.tmp" "${output}"
    VERBATIM
    MAIN_DEPENDENCY "${source}"
)
```

File: Meta/CMake/serenity_components.cmake [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    endif()
+    if(BUILD_EVERYTHING OR BUILD_${NAME_UPPER} OR SERENITY_COMPONENT_REQUIRED)
+        add_dependencies(components "Component${name}")
+    endif()
+    foreach(dependency ${SERENITY_COMPONENT_DEPENDS})
+        add_dependencies("Component${name}" "Component${dependency}")
+    endforeach()
+endfunction()
+
+macro(export_components_helper file_name current_dir)
```
[the remark this answers is not in the document]
Comment (ADKaster) [reply Meta/CMake/serenity_components.cmake:35]: Well, it's not quite true that it's only a move, I also changed the set_/get_property(TARGET target PROPERTY ...) calls into set_/get_target_property(). So This one seems reasonable enough to clean up while we're here.

File: CMakeLists.txt (hunk 2 above)
[the remark this answers is not in the document]
Comment (BenWiederhake) [reply CMakeLists.txt:322]: The reason for `write-only-on-difference.sh` is that I didn't know that `copy_if_different` exists :D

If this were for many thousands of files, I'd worry whether this approach is measurably slower, because `write-only-on-difference.sh` might cache more efficiently. However, that's probably a non-issue.

[the remark this answers is not in the document]
Comment (ADKaster) [reply CMakeLists.txt:322]: Well... One difference `write-only-on-difference.sh` *does* have is that it cleans up all those byproduct files after it creates them. And if we look at, for example, all the JavaScript wrapper bindings we generate...
```
:~/serenity/Build/i686$ find ./Userland/Libraries/LibWeb -name "*.tmp" | wc -l
672
```
There sure are a lot of them 😅 .

```
~/serenity/Build/i686$ find . -name "*.tmp" | wc -l
753
```

Looking at 753 `.tmp` headers/cpp files laying around in your build directory once we add in the gml files and the IPC files. Which I have no idea if the filesystem usage there at rest matters (probably not? idk) or if the fact that we are overwriting tmp files rather than regenerating them entirely for partial rebuilds matters more.

[the remark this answers is not in the document]
Comment (ADKaster) [reply CMakeLists.txt:322]:
```
 find . -name "*.tmp" -print0 | du --files0-from=- -ch | tail -1
4.8M    total
```

5 Megabytes? meh.

Though there are some noticeable chonkers in there:
```
40K     ./Userland/Services/LaunchServer/LaunchServerEndpoint.h.tmp
52K     ./Userland/Services/AudioServer/AudioServerEndpoint.h.tmp
64K     ./Userland/Services/WebContent/WebContentServerEndpoint.h.tmp
68K     ./Userland/Libraries/LibWeb/Bindings/HTMLElementPrototype.cpp.tmp
88K     ./Userland/Libraries/LibWeb/Bindings/DocumentPrototype.cpp.tmp
88K     ./Userland/Services/WebContent/WebContentClientEndpoint.h.tmp
100K    ./Userland/Services/WindowServer/WindowClientEndpoint.h.tmp
184K    ./Userland/Libraries/LibVT/EscapeSequenceStateMachine.h.tmp
284K    ./Userland/Services/WindowServer/WindowServerEndpoint.h.tmp
```

[the remark this answers is not in the document]
Comment (BenWiederhake) [reply CMakeLists.txt:322]: It probably doesn't make a big difference. But if you really care, maybe it can be cleaned up by an additional
`COMMAND rm "${output}.tmp"` or something?

[the remark this answers is not in the document]
Comment (friendlyanon) [reply CMakeLists.txt:322]: Yeah, it's not a requirement to keep byproducts around. Removal should preferably happen with `"${CMAKE_COMMAND}" -E remove "${output}.tmp"`.

File: CMakeLists.txt [hunk 4]
```diff
@@ -14,6 +14,11 @@ if (CMAKE_COMPILER_IS_GNUCXX AND CMAKE_CXX_COMPILER_VERSION VERSION_LESS 10.2)
       "Please re-read the build instructions documentation, and upgrade your host compiler.\n")
 endif()
 
+get_filename_component(
+    SERENITY_PROJECT_ROOT "${PROJECT_SOURCE_DIR}"
+    ABSOLUTE CACHE
+)
+
```
[the remark this answers is not in the document]
Comment (ADKaster) [reply CMakeLists.txt:21]: Reason I did it was the problem child wasm_spec_tests.cmake which is called from both the Lagom and SerenityOS projects. But, that script needs some extra help anyway, as it currently downloads files into the binary dir, and generates files into the source tree. So I don't mind adding another quick hack to it to let it do ... whatever it wants to do a bit longer.

File: Meta/CMake/utils.cmake (hunk 1 above)
Comment (ADKaster) [reply Meta/CMake/utils.cmake:3]: When I make that change, it complains that it can't find the files
```
CMake Error at Meta/CMake/utils.cmake:2 (include):
  include could not find load file:

    serenity_components.cmake
Call Stack (most recent call first):
  CMakeLists.txt:326 (include)
```
Even when utils is included like
```
include(Meta/CMake/utils.cmake)
```
Probably related to not adding Meta/CMake to the CMake include path 🤷‍♂️

Comment (friendlyanon) [reply Meta/CMake/utils.cmake:3]: Right, `include`d things are dynamically scoped, which applies to paths as well. This can be resolved then.

File: Userland/Libraries/LibWeb/CMakeLists.txt [hunk 5]
```diff
@@ -242,70 +242,54 @@ set(GENERATED_SOURCES
...
+    math(EXPR bindings_end "${num_bindings} - 1")
+    foreach(iter RANGE "${bindings_end}")
+        list(GET BINDINGS_SOURCES ${iter} bindings_src)
+        list(GET BINDINGS_TYPES ${iter} bindings_type)
+        add_custom_command(
+            OUTPUT "${bindings_src}"
+            BYPRODUCTS "${bindings_src}.tmp"
+            COMMAND "$<TARGET_FILE:WrapperGenerator>" "--${bindings_type}" "${CMAKE_CURRENT_SOURCE_DIR}/${class}.idl" > "${bindings_src}.tmp"
+            COMMAND "${CMAKE_COMMAND}" -E copy_if_different "${bindings_src}.tmp" "${bindings_src}"
+            COMMAND "${CMAKE_COMMAND}" -E rm "${bindings_src}.tmp"
```
Comment (friendlyanon) [inline Userland/Libraries/LibWeb/CMakeLists.txt:282]: `rm` was introduced in 3.17, but the lists files require 3.16 as a minimum. `remove` should be used instead.

**Review by friendlyanon — CHANGES_REQUESTED** [review]

File: CMakeLists.txt [hunk 6]
```diff
@@ -72,37 +75,37 @@ set(GCC_VERSION 11.2.0)
...
 add_custom_target(install-ports
-    COMMAND ${CMAKE_COMMAND} -E env "SERENITY_SOURCE_DIR=${CMAKE_SOURCE_DIR}" "SERENITY_ARCH=${SERENITY_ARCH}" ${CMAKE_SOURCE_DIR}/Meta/install-ports-tree.sh
+    COMMAND ${CMAKE_COMMAND} -E env "SERENITY_SOURCE_DIR=${SERENITY_PROJECT_ROOT}" "SERENITY_ARCH=${SERENITY_ARCH}" ${SERENITY_PROJECT_ROOT}/Meta/install-ports-tree.sh
     USES_TERMINAL
 )
 
 add_custom_target(configure-components
-    COMMAND ConfigureComponents
+    COMMAND $<TARGET_FILE:ConfigureComponents>
     DEPENDS ConfigureComponents
```
[the remark this answers is not in the document]
Comment (ADKaster) [reply CMakeLists.txt]: Aha makes sense. I removed a majority of the bogus DEPENDS, but it's possible I misunderstood the rules about files in the source tree, e.g. some of the shell scripts we use in these custom command/custom target commands.

File: Userland/Libraries/LibWeb/CMakeLists.txt (hunk 5 above)
Comment (ADKaster) [reply Userland/Libraries/LibWeb/CMakeLists.txt:282]: D'oh. Didn't read the docs carefully enough it seems. Just saw that `rm` was "preferred".

File: CMakeLists.txt (hunk 6 above)
[the remark this answers is not in the document]
Comment (friendlyanon) [reply CMakeLists.txt]: Only [add_custom_command](https://cmake.org/cmake/help/latest/command/add_custom_command.html) can establish target-level dependency with its `DEPENDS` argument according to the documentation. You should also think about what is a concern during configure time and build time. For example, I have the generator script that's being invoked in this [example](https://github.com/friendlyanon/cmake-init-gif-engine/blob/master/CMakeLists.txt#L29-L30) listed as a dependency, such that when that script is updated, I get proper dependency tracking. Something like this was also missing for that write_if_different script that was in use before.

[the remark this answers is not in the document]
Comment (ADKaster) [reply CMakeLists.txt]: Hmm so, it's incorrect to remove DEPENDS IPCCompiler from this custom command, because the target-level dependency added by having `$<TARGET_FILE:IPCCompiler>`  in a COMMAND statement doesn't cause the command to be re-run when we re-compile IPCCompiler?

```
    add_custom_command(
        OUTPUT ${output}
        BYPRODUCTS ${output}.tmp
        COMMAND $<TARGET_FILE:IPCCompiler> ${source} > ${output}.tmp
        COMMAND "${CMAKE_COMMAND}" -E copy_if_different ${output}.tmp ${output}
        COMMAND "${CMAKE_COMMAND}" -E remove ${output}.tmp
-       DEPENDS IPCCompiler
        VERBATIM
        MAIN_DEPENDENCY ${source}
    )
```

> This target-level dependency does NOT add a file-level dependency that would cause the custom command to re-run whenever the executable is recompiled. List target names with the DEPENDS option to add such file-level dependencies.

And for the add_custom_target, we should `add_dependencies(configure-components ConfigureComponents)` ?

[the remark this answers is not in the document]
Comment (friendlyanon) [reply CMakeLists.txt]: I saw you put a FIXME for `ZIP_LISTS`, it might be worth putting one here as well, since `$<TARGET_FILE:...>` does not establish a target-level dependency if [CMP0112](https://cmake.org/cmake/help/latest/policy/CMP0112.html) is set to `NEW` (i.e. when a version >=3.19 is requested). I don't know if Serenity will ever have a reason to do that, but it might be worth keeping in mind, even if you consider incorrect parallel builds for ExternalProject before 3.19, for which there is an [easy fix](https://github.com/friendlyanon/cmake-init-gif-engine/commit/662ca8894d833c8c945c537c5d97d95dc6ffa4cf).

[the remark this answers is not in the document]
Comment (ADKaster) [reply CMakeLists.txt]: When I read the documentation on those target-level dependencies I read it as, "TARGET_FILE, TARGET_LINKER_FILE, TARGET_SONAME_FILE, and TARGET_PDB_FILE establish a target level dependency, but with CMP0112 set to NEW, these extended generator expressions with _NAME, _DIR, and _BASE_NAME will stop adding the target dependency". I take it this means that TARGET_FILE actually is some combo of TARGET_FILE_NAME and TARGET_FILE_DIR?

There's a large chunk of refactoring left to get the ExternalProject stuff up and going to have proper CMakeToolchain support for the target build.

Regarding upgrading versions, I asked around in our discord and the consensus was that until there's some killer feature that we can't do without, it's preferred to stay on a CMake version supported by the current major LTS releases (Ubuntu 20.04 seems to be sticking with 3.16 through the ubuntu PPAs).

> I guess this already sort of answers your question.

Also yeah, pretty much. I'll be adding the dependencies manually in a push soon ™️ .

File: Meta/CMake/serenity_components.cmake [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+#
+
+define_property(TARGET PROPERTY SERENITY_COMPONENT_NAME BRIEF_DOCS "SerenityOS component name" FULL_DOCS "-")
+define_property(TARGET PROPERTY SERENITY_COMPONENT_DESCRIPTION BRIEF_DOCS "SerenityOS component description" FULL_DOCS "-")
+define_property(TARGET PROPERTY SERENITY_COMPONENT_RECOMMENDED BRIEF_DOCS "SerenityOS component recommended (flag)" FULL_DOCS "-")
+define_property(TARGET PROPERTY SERENITY_COMPONENT_REQUIRED BRIEF_DOCS "SerenityOS component required (flag)" FULL_DOCS "-")
+define_property(TARGET PROPERTY SERENITY_COMPONENT_DEPENDS BRIEF_DOCS "SerenityOS component dependencies" FULL_DOCS "-")
+
+function(serenity_component name)
+    cmake_parse_arguments(SERENITY_COMPONENT "RECOMMENDED;REQUIRED" "DESCRIPTION" "TARGETS;DEPENDS" ${ARGN})
```
[the remark this answers is not in the document]
Comment (ADKaster) [reply Meta/CMake/serenity_components.cmake]: Sounds good, I've moved our uses of cmake_parse_arguments to that, except for the ones in Lagom. Plan to revisit those guys later to move lagom_lib and lagom_test to a helper file...

File: Userland/Libraries/LibWeb/CMakeLists.txt [hunk 8]
```diff
@@ -242,70 +242,51 @@ set(GENERATED_SOURCES
...
+    math(EXPR bindings_end "${num_bindings} - 1")
+    foreach(iter RANGE "${bindings_end}")
+        list(GET BINDINGS_SOURCES ${iter} bindings_src)
+        list(GET BINDINGS_TYPES ${iter} bindings_type)
+        add_custom_command(
+            OUTPUT "${bindings_src}"
+            BYPRODUCTS "${bindings_src}.tmp"
+            COMMAND "$<TARGET_FILE:WrapperGenerator>" "--${bindings_type}" "${CMAKE_CURRENT_SOURCE_DIR}/${class}.idl" > "${bindings_src}.tmp"
+            COMMAND "${CMAKE_COMMAND}" -E copy_if_different "${bindings_src}.tmp" "${bindings_src}"
+            COMMAND "${CMAKE_COMMAND}" -E remove "${bindings_src}.tmp"
```
Comment (ADKaster) [inline Userland/Libraries/LibWeb/CMakeLists.txt:278]: Ok now I'm confused. What is it about the change to using the copy_if_different and remove that causes these commands to run every single time you invoke `ninja` or `ninja install` ? Kind of a bummer to create 753 files  just to do no work in the end...

Comment (friendlyanon) [reply Userland/Libraries/LibWeb/CMakeLists.txt:278]: You still have the `BYPRODUCTS` there. The build system sees that the byproducts are missing and so regenerates them.

Comment (ADKaster) [reply Userland/Libraries/LibWeb/CMakeLists.txt:277]: Oh! That makes a lot of sense then. I've removed that now 🤦‍♂️ .

File: Meta/CMake/serenity_components.cmake [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+        SERENITY_COMPONENT_RECOMMENDED "${SERENITY_COMPONENT_RECOMMENDED}"
+        SERENITY_COMPONENT_REQUIRED    "${SERENITY_COMPONENT_REQUIRED}"
+        SERENITY_COMPONENT_DEPENDS     "${SERENITY_COMPONENT_DEPENDS}"
+    )
+    if(NOT "${SERENITY_COMPONENT_TARGETS}" STREQUAL "") # note: one component is /bin/false, which makes this tricky
+        foreach(target IN LISTS SERENITY_COMPONENT_TARGETS)
+            add_dependencies("Component${name}" "${target}")
+        endforeach()
+    endif()
+    if(BUILD_EVERYTHING OR "BUILD_${NAME_UPPER}" OR SERENITY_COMPONENT_REQUIRED)
```
Comment (friendlyanon) [inline Meta/CMake/serenity_components.cmake:29]: I just noticed, but this should be dereferenced here. See [CMP0054](https://cmake.org/cmake/help/latest/policy/CMP0054.html)
```suggestion
    if(BUILD_EVERYTHING OR "${BUILD_${NAME_UPPER}}" OR SERENITY_COMPONENT_REQUIRED)
```

File: CMakeLists.txt [hunk 10]
```diff
@@ -137,30 +135,22 @@ if (ENABLE_ALL_DEBUG_FACILITIES)
...
 configure_file(Kernel/Debug.h.in Kernel/Debug.h @ONLY)
 
+include(Meta/CMake/wasm_spec_tests.cmake)
+
 include_directories(Userland/Libraries)
 include_directories(.)
 include_directories(${CMAKE_BINARY_DIR})
 
-set(write_if_different ${CMAKE_SOURCE_DIR}/Meta/write-only-on-difference.sh)
+set(write_if_different ${SerenityOS_SOURCE_DIR}/Meta/write-only-on-difference.sh)
```
Comment (friendlyanon) [inline CMakeLists.txt:150]: This script was deleted. This variable should be removed as well.

**Review by friendlyanon — CHANGES_REQUESTED** [review]

File: CMakeLists.txt (hunk 10 above)
Comment (ADKaster) [reply CMakeLists.txt:150]: The variable is deleted after the commit that makes this change, in the last commit of the PR (after all callers are refactored)
The commit that deletes the script removes the variable.

Comment (friendlyanon) [reply CMakeLists.txt:150]: It's changed in https://github.com/SerenityOS/serenity/pull/9281/commits/af2c78c3df835bbb644285c7a82596f8ca7efdb7 but missing from https://github.com/SerenityOS/serenity/pull/9281/commits/d15f00066fd814f702c509590ce9a71a0b8c99d7

File: Meta/CMake/serenity_components.cmake (hunk 9 above)
Comment (friendlyanon) [reply Meta/CMake/serenity_components.cmake:29]: Same on line 69 (nice) as well. See [if()](https://cmake.org/cmake/help/latest/command/if.html#basic-expressions) for how these would be evaluated.

Comment (ADKaster) [reply Meta/CMake/serenity_components.cmake]: oh! good catch. Yeah we want to check these vs "ON/OFF" rather than vs existence.

File: CMakeLists.txt (hunk 10 above)
Comment (ADKaster) [reply CMakeLists.txt]: All right. I've modified the history so that the only commit that touches this variable is the one that deletes it.

File: Meta/CMake/code_generators.cmake [hunk 11]
```diff
@@ -0,0 +1,55 @@
+#
+# Functions for generating sources using host tools
+#
+
+function(compile_gml source output string_name)
+    set(source ${CMAKE_CURRENT_SOURCE_DIR}/${source})
+    add_custom_command(
+        OUTPUT ${output}
+        COMMAND ${SerenityOS_SOURCE_DIR}/Meta/text-to-cpp-string.sh ${string_name} ${source} > ${output}.tmp
```
Comment (friendlyanon) [inline Meta/CMake/code_generators.cmake:9]: Technically, this could be replaced with a CMake script to make this less platform dependent. This script is fairly trivial as well:
```cmake
# Inputs: ${input} - path to file to read, ${output} - path to output to, ${string_name} - variable name
cmake_minimum_required(VERSION 3.16)

file(READ "${input}" content)

# Escape semicolons, so they are preserved as-is
string(REPLACE ";" "\\;" content "${content}")
# Turn each line into a list element
string(REPLACE "\n" ";" content "${content}")

set(result "\
extern const char ${string_name}[];
const char ${string_name}[] = R\"~~~(
")

foreach(line IN LISTS content)
  if(NOT line STREQUAL "" AND NOT line MATCHES "^ *#")
    string(APPEND result "${line}\n")
  endif()
endforeach()

string(APPEND result ")~~~\";\n")

file(WRITE "${output}" "${result}")
```

Comment (ADKaster) [reply Meta/CMake/code_generators.cmake:9]: Hrm. I'm not sure that eliminating the host platform dependency of

- having bash, and `/usr/bin/env`
- having grep that supports the regex '^ *#'

is of immediate concern to the project? 

I think until we figure out a reasonable way to create our cross-compilers for a Windows or non-posix host, (and have actual users/maintainers for such platforms), we'll pass on this particular cmake-ism.. the shell script is like 10 lines, and this script is about the same length as the current solution. Changing BuildIt.sh and BuildClang.sh to be platform independent would be a more interesting platform independence project imo 🤷‍♂️ .

Comment (friendlyanon) [reply Meta/CMake/code_generators.cmake:9]: Yes, it was just a comment that I had to get out of my system. Technically, there is nothing that prevents Serenity from being compiled on platforms other than *nix.

Comment (ADKaster) [reply Meta/CMake/code_generators.cmake:9]: I know that feel. Also true, nothing preventing it assuming one can create the cross-compiler. Especially once I figure out the proper toolchain setup in my superbuild branch. Speaking of which, are we good on this PR? Feels like it's bout that time to get this in and iterate further on some future PRs.

**Review by friendlyanon — APPROVED** [review]
LGTM!

**Comment by ADKaster** [comment]
D'oh forgot about the CI cache of the unicode data cmake file, so fixed that in this latest rebase + push.

File: Meta/Lagom/Tools/CodeGenerators/StateMachineGenerator/main.cpp [hunk 12]
```diff
@@ -252,7 +252,7 @@ HashTable<String> actions(const StateMachine& machine)
     }
     if (machine.anywhere.has_value())
         do_state(machine.anywhere.value());
-    return move(table);
+    return table;
```
Comment (BenWiederhake) [inline Meta/Lagom/Tools/CodeGenerators/StateMachineGenerator/main.cpp:255]: This probably belongs into a different commit? I have no idea what it does, but it's probably fine.

File: Meta/write-only-on-difference.sh [hunk 13]
```diff
@@ -1,28 +0,0 @@
-#!/bin/sh
```
Comment (BenWiederhake) [inline Meta/write-only-on-difference.sh:1]: /me waves goodbye

It's good to see that it is being replaced by a proper pointless-rebuild-avoidance-mechanism. :)

**Review by BenWiederhake — COMMENTED** [review]
I'm so far behind that I don't really know what half the touched code is about. I didn't even know we're using UCD now, awesome! :D
(Also: Damn my broken computer.)

See two minor comments below, but I'm definitely not the right person to do an actual review. It looks like a good idea though! :)

File: Meta/Lagom/Tools/CodeGenerators/StateMachineGenerator/main.cpp (hunk 12 above)
Comment (ADKaster) [reply Meta/Lagom/Tools/CodeGenerators/StateMachineGenerator/main.cpp:255]: It prevents a -Wpessimizing-move warning when built with all the proper warnings. before this commit, the Code Generators were built with less warnings than the rest of the system. but fair, removing a move is a bit of a different change to a file move.

**Review by linusg — APPROVED** [review]
rad :)
